psi4/driver/p4util/python_helpers.py

```diff
@@ -123,6 +123,120 @@ def pybuild_wavefunction(mol, basis=None):
+
+    for label in wfn_dimension:
+        tup = wfn_dimension[label]
+        wfn_dimension[label] = core.Dimension.from_list(tup,name=label) if tup else None
+
+    wfn = core.Wavefunction(molecule, basisset, wfn_matrix, wfn_vector, wfn_dimension, wfn_int,
+                            wfn_string, wfn_boolean, wfn_float)
+    return wfn
+
+core.Wavefunction.from_file = pyread_wavefunction
+
+
+@staticmethod
+def pywrite_wavefunction(wfn, filename):
+    # collect the wavefunction's variables in a dictionary indexed by varaible type
+    # some of the data types have to be made numpy-friendly first
+    wfn_data = {
+        'molecule' : wfn.molecule().to_dict(),
+        'matrix' : {
+            'Ca'       : wfn.Ca().to_array()       if wfn.Ca()       else None,
+            'Cb'       : wfn.Cb().to_array()       if wfn.Cb()       else None,
+            'Da'       : wfn.Da().to_array()       if wfn.Da()       else None,
+            'Db'       : wfn.Db().to_array()       if wfn.Db()       else None,
+            'Fa'       : wfn.Fa().to_array()       if wfn.Fa()       else None,
+            'Fb'       : wfn.Fb().to_array()       if wfn.Fb()       else None,
+            'H'        : wfn.H().to_array()        if wfn.H()        else None,
+            'S'        : wfn.S().to_array()        if wfn.Ca()       else None,
+            'X'        : wfn.X().to_array()        if wfn.X()        else None,
+            'aotoso'   : wfn.aotoso().to_array()   if wfn.aotoso()   else None,
+            'gradient' : wfn.gradient().to_array() if wfn.gradient() else None,
+            'hessian'  : wfn.hessian().to_array()  if wfn.hessian()  else None
+            },
+        'vector' : {
+            'epsilon_a'            : wfn.epsilon_a().to_array()   if wfn.epsilon_a() else None,
+            'epsilon_b'            : wfn.epsilon_b().to_array()   if wfn.epsilon_b() else None,
+            'frequencies'          : wfn.frequencies().to_array() if wfn.frequencies() else None
+            #'atomic_point_charges' : wfn.atomic_point_charges().to_array() if wfn.atomic_point_charges() else None,
+            #'esp_at_nuclei'        : wfn.esp_at_nuclei().to_array() if wfn.esp_at_nuclei() else None,
+            },
+        'dimension' : {
+            'doccpi'   : wfn.doccpi().to_tuple(),
+            'frzcpi'   : wfn.frzcpi().to_tuple(),  
+            'frzvpi'   : wfn.frzvpi().to_tuple(),  
+            'nalphapi' : wfn.nalphapi().to_tuple(),
+            'nbetapi'  : wfn.nbetapi().to_tuple(), 
+            'nmopi'    : wfn.nmopi().to_tuple(),   
+            'nsopi'    : wfn.nsopi().to_tuple(),   
+            'soccpi'   : wfn.soccpi().to_tuple()  
+            },
+        'int' : { 
+            'nalpha' : wfn.nalpha(),
+            'nbeta'  : wfn.nbeta(),
+            'nfrzc'  : wfn.nfrzc(),
+            'nirrep' : wfn.nirrep(),
+            'nmo'    : wfn.nmo(),
+            'nso'    : wfn.nso()
```

Comment:
`print_lvl_`,

Reply:
Will probably be important to decide and enforce what the minimal expected fields are for a serialized Wfn. There's some "dummy wfn"s out there with only mol & basis set contents.

Reply:
mol & basis wavefunctions should initialize all of the above IIRC, if not we might want to consider doing so.